src/tasks/gitea/gitea-drone-oauth.ts

```diff
@@ -0,0 +1,212 @@
+import * as k8s from '@kubernetes/client-node'
+import axios, { AxiosRequestConfig } from 'axios'
+import querystring from 'querystring'
+
+import { UserApi, CreateOAuth2ApplicationOptions, OAuth2Application } from '@redkubes/gitea-client-node'
+
+import { cleanEnv, GITEA_USER, GITEA_PASSWORD, GITEA_URL, GITEA_REPO } from '../../validators'
+
+const env = cleanEnv({
+  GITEA_USER,
+  GITEA_PASSWORD,
+  GITEA_URL,
+  GITEA_REPO,
+})
+type DroneSecret = {
+  clientId: string
+  clientSecret: string
+}
+class GiteaDroneOAuth {
+  apiClient: k8s.CoreV1Api
+  k8s_namespace = 'gitea'
+  k8s_secret_name = 'gitea-drone-secret'
+  csrfCookieName = '_csrf'
+
+  oauthData: DroneSecret
+
+  giteaUrl: string
+  droneUrl: string
+
+  csrfToken: string
+  constructor() {
+    this.giteaUrl = env.GITEA_URL
+    if (this.giteaUrl.endsWith('/')) {
+      this.giteaUrl = this.giteaUrl.slice(0, -1)
+    }
+    this.droneUrl = this.giteaUrl.replace('gitea', 'drone')
+
+    this.main()
+  }
+
+  getApiClient(): k8s.CoreV1Api {
+    if (this.apiClient) return this.apiClient
+    const kc = new k8s.KubeConfig()
+    kc.loadFromDefault()
+    this.apiClient = kc.makeApiClient(k8s.CoreV1Api)
+    return this.apiClient
+  }
+
+  async secretExists(): Promise<DroneSecret> {
+    const client = this.getApiClient()
+    try {
+      const response = (await client.readNamespacedSecret(this.k8s_secret_name, this.k8s_namespace)).body
+      return response.data as DroneSecret
+    } catch (e) {
+      return undefined
+    }
+  }
+  async authorizeOAuth() {
+    const options: AxiosRequestConfig = {
+      params: {
+        // eslint-disable-next-line @typescript-eslint/camelcase
+        client_id: this.oauthData.clientId,
+        // eslint-disable-next-line @typescript-eslint/camelcase
+        redirect_uri: `${this.droneUrl}/login`,
+        // eslint-disable-next-line @typescript-eslint/camelcase
+        response_type: 'code',
+      },
+      maxRedirects: 1,
+      auth: {
+        username: env.GITEA_USER,
+        password: env.GITEA_PASSWORD,
+      },
+    }
+
+    console.log('Authorizing OAuth application')
+
+    let authorizeResponse
+    try {
+      authorizeResponse = await axios.get(`${this.giteaUrl}/login/oauth/authorize`, options)
+    } catch (error) {
+      console.log('Authorization already granted or something went wrong')
```

Comment:
Is it possible to distinguish error from `Authorization already granted ` by inspecting HTTP repose code ?

Reply:
No sadly not, it always returns the same HTTP response code in case of an error. And I found, that it is reasonable to assume that authorization was already granted

Reply:
The HTTP code is 400 - Bad Request, no more information is given

Reply:
ok, then I do not have any remarks here. Thanks